File: src/main/java/com/mercateo/jsonschema/property/Property.java
package com.mercateo.jsonschema.property;

import com.mercateo.immutables.Tuple;
import com.mercateo.jsonschema.generictype.GenericType;
import javaslang.collection.List;
import javaslang.collection.Map;
import javaslang.collection.Set;
import org.immutables.value.Value;

import java.lang.annotation.Annotation;
import java.util.Collection;
import java.util.function.Function;

@Value.Immutable
@Tuple
public interface Property {
    String name();

    PropertyDescriptor propertyDescriptor();

    Function valueAccessor();

    default Object getValue(Object object) {
       return valueAccessor().apply(object);
    }

    default List<Property> children() {
        return propertyDescriptor().children();
    }

    default GenericType<?, ?> genericType() {
        return propertyDescriptor().genericType();
    }

    Map<Class<? extends Annotation>, Set<Annotation>> annotations();
}

File: src/main/java/com/mercateo/jsonschema/property/Property.kt
package com.mercateo.jsonschema.property

import com.mercateo.jsonschema.generictype.GenericType

data class Property<in S, T>(
        val name: String,
        val propertyDescriptor: PropertyDescriptor<T>,
        val valueAccessor: (S) -> T?,
        val annotations: Map<Class<out Annotation>, Set<Annotation>>,
        val context: Context = Property.Context.Unconnected
) {
    fun getValue(instance: S): T? {
        return valueAccessor(instance)
    }

    fun filterChildren(predicate: (Property<Any, Any>) -> Boolean): Property<S, T> {
        return copy(propertyDescriptor= propertyDescriptor.filter(predicate))
    }

    fun update(updateVisitor: PropertyDescriptor.UpdateVisitor) : Property<S, T> {
        return updateVisitor.update(this as Property<Any, Any>) as Property<S, T>
    }

    fun updateChildren(updateVisitor: PropertyDescriptor.UpdateVisitor) : Property<S, T> {
        return copy(propertyDescriptor = propertyDescriptor.update(updateVisitor))
    }

    val children: List<Property<T, Any>> = propertyDescriptor.children

    val reference: String? = context.reference

    val path: String? = context.path

    val genericType: GenericType<T> = propertyDescriptor.genericType

    val propertyType: PropertyType = propertyDescriptor.propertyType

    sealed class Context(val path: String? = null, val reference: String? = null) {
        object Unconnected : Context()

        class Connected(path: String, reference: String?) : Context(path, reference) {
            override fun toString(): String {
                return "Connected(path='$path', reference='$reference')"
            }
        }
    }
}
